chain/index/chain_indexer.go

```diff
@@ -0,0 +1,723 @@
+	evs []*types.Event
+}
+
+func (ci *ChainIndexer) loadExecutedMessages(ctx context.Context, msgTs, rctTs *types.TipSet) ([]executedMessage, error) {
+	msgs, err := ci.cs.MessagesForTipset(ctx, msgTs)
+	if err != nil {
+		return nil, xerrors.Errorf("read messages: %w", err)
+	}
+
+	st := ci.cs.ActorStore(ctx)
+
+	arr, err := blockadt.AsArray(st, rctTs.Blocks()[0].ParentMessageReceipts)
+	if err != nil {
+		return nil, xerrors.Errorf("load receipts amt: %w", err)
+	}
+
+	if uint64(len(msgs)) != arr.Length() {
+		return nil, xerrors.Errorf("mismatching message and receipt counts (%d msgs, %d rcts)", len(msgs), arr.Length())
+	}
+
+	ems := make([]executedMessage, len(msgs))
+
+	for i := 0; i < len(msgs); i++ {
+		ems[i].msg = msgs[i]
+
+		var rct types.MessageReceipt
+		found, err := arr.Get(uint64(i), &rct)
+		if err != nil {
+			return nil, xerrors.Errorf("load receipt: %w", err)
+		}
+		if !found {
+			return nil, xerrors.Errorf("receipt %d not found", i)
+		}
+		ems[i].rct = &rct
+
+		if rct.EventsRoot == nil {
+			continue
+		}
+
+		evtArr, err := amt4.LoadAMT(ctx, st, *rct.EventsRoot, amt4.UseTreeBitWidth(types.EventAMTBitwidth))
+		if err != nil {
+			return nil, xerrors.Errorf("load events amt: %w", err)
+		}
+
+		ems[i].evs = make([]*types.Event, evtArr.Len())
+		var evt types.Event
+		err = evtArr.ForEach(ctx, func(u uint64, deferred *cbg.Deferred) error {
+			if u > math.MaxInt {
+				return xerrors.Errorf("too many events")
+			}
+			if err := evt.UnmarshalCBOR(bytes.NewReader(deferred.Raw)); err != nil {
+				return err
+			}
+
+			cpy := evt
+			ems[i].evs[int(u)] = &cpy //nolint:scopelint
+			return nil
+		})
+
+		if err != nil {
+			return nil, xerrors.Errorf("read events: %w", err)
+		}
+
+	}
+
+	return ems, nil
+}
+
+func (ci *ChainIndexer) GetMsgInfo(ctx context.Context, msg_cid cid.Cid) (MsgInfo, error) {
+	msgInfo, err := ci.lookupMsg(ctx, msg_cid)
+	if err == nil {
+		return msgInfo, nil
+	}
+
+	if err == sql.ErrNoRows {
+		// wait till head is processed and retry
+		if err := ci.waitTillHeadIndexed(ctx); err != nil {
+			return MsgInfo{}, err
+		}
+		msgInfo, err := ci.lookupMsg(ctx, msg_cid)
+		if err == sql.ErrNoRows {
+			return MsgInfo{}, ErrNotFound
+		}
+		return msgInfo, err
+	}
+	return MsgInfo{}, err
+}
+
+func (ci *ChainIndexer) lookupMsg(ctx context.Context, msg_cid cid.Cid) (MsgInfo, error) {
+	var tipsetKeyCidBytes []byte
+	var height uint64
+
+	row := ci.stmts.stmtSelectMsg.QueryRowContext(ctx, msg_cid.Bytes())
+	err := row.Scan(&tipsetKeyCidBytes, &height)
+	if err != nil {
+		return MsgInfo{}, err
+	}
+
+	tsKeyCid, err := cid.Cast(tipsetKeyCidBytes)
+	if err != nil {
+		return MsgInfo{}, xerrors.Errorf("error casting tipset key cid: %w", err)
+	}
+
+	return MsgInfo{
+		Message: msg_cid,
+		Epoch:   abi.ChainEpoch(height),
+		TipSet:  tsKeyCid,
+	}, nil
+}
+
+func (ci *ChainIndexer) waitTillHeadIndexed(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
```

Comment:
should we just use `EpochDurationSeconds` here?

Reply:
@rvagg Honestly, this has nothing to do with an epoch. I think we can change the timeout to 5seconds. The idea is that the Index should reflect the heaviest block we have in our statestore that's it. We're not waiting for a new block to come in here.